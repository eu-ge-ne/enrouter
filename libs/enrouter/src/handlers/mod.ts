import type { ReactElement } from "react";
import * as regexparam from "regexparam";

import { createLog } from "#log.js";

import type { Route } from "#route/mod.js";

const log = createLog("handlers");

export interface RouteHandler {
  route: Route;

  /**
   * @see https://github.com/lukeed/regexparam
   */
  test: { keys: string[]; pattern: RegExp };

  modules: { id: string; loaded?: true }[];

  layout?: Record<string, ReactElement>;
  index?: Record<string, ReactElement>;
  notFound?: Record<string, ReactElement>;

  tree?: RouteHandler[];
}

interface BuildRouteHandlersParams {
  routes: Route;
}

export function buildRouteHandlers({
  routes,
}: BuildRouteHandlersParams): RouteHandler {
  log("Building route handlers");

  const handlers = recur(routes);

  log("Route handlers built: %O", handlers);

  return handlers;
}

function recur(route: Route): RouteHandler {
  const handler: RouteHandler = {
    route,
    test: regexparam.parse(route.path, true),
    modules: route.mod.map((id) => ({ id })),
  };

  if (route.tree) {
    handler.tree = route.tree.map(recur);
  }

  return handler;
}
