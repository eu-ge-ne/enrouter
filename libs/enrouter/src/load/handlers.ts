import { createLog } from "#log.js";
import { loaders } from "./loaders.js";

import type { RouteModules } from "#modules.js";
import type { RouteHandler } from "#handlers/mod.js";

const log = createLog("load/handlers");

interface LoadRouteHandlersParams {
  handlers: RouteHandler;
  modules: RouteModules;
}

export async function loadRouteHandlers({
  handlers,
  modules,
}: LoadRouteHandlersParams): Promise<void> {
  log("Loading route handlers");

  const promises: Promise<unknown>[] = [];

  recur([handlers], modules, promises);

  await Promise.all(promises);

  log("Route handlers loaded");
}

function recur(
  handlers: RouteHandler[],
  modules: RouteModules,
  promises: Promise<unknown>[],
): void {
  for (const handler of handlers) {
    for (const module of handler.modules) {
      if (module.loaded) {
        continue;
      }

      const { path, load } = modules[module.id]!;

      const fileName = path.slice(path.lastIndexOf("/") + 1);

      const promise = loaders[fileName]?.({ handler, module, load });
      if (promise) {
        promises.push(promise);
      }
    }

    if (handler.tree) {
      recur(handler.tree, modules, promises);
    }
  }
}
