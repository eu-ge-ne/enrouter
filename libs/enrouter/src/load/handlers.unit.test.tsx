import { describe, test, expect } from "vitest";
import * as regexparam from "regexparam";

import { loadRouteHandlers } from "./handlers.js";

import type { RouteModules } from "#modules.js";
import type { RouteHandler } from "#handlers/mod.js";

describe("loadRouteHandlers", () => {
  test("1 route handler with no modules", async () => {
    const handlers: RouteHandler = {
      route: {
        path: "/",
        mod: [],
        link: { css: [], mod: [] },
      },
      test: regexparam.parse("/", true),
      modules: [],
    };

    await loadRouteHandlers({ handlers, modules: {} });

    expect(handlers).toMatchSnapshot();
  });

  test("2 route handlers with no modules", async () => {
    const handlers: RouteHandler = {
      route: {
        path: "/",
        mod: [],
        link: { css: [], mod: [] },
      },
      test: regexparam.parse("/", true),
      modules: [],
      children: [
        {
          route: {
            path: "/abc",
            mod: [],
            link: { css: [], mod: [] },
          },
          test: regexparam.parse("/abc", true),
          modules: [],
        },
      ],
    };

    await loadRouteHandlers({ handlers, modules: {} });

    expect(handlers).toMatchSnapshot();
  });

  test("1 route handler with 1 _layout.tsx module", async () => {
    const handlers: RouteHandler = {
      route: {
        path: "/",
        mod: ["_layout.tsx"],
        link: { css: [], mod: [] },
      },
      test: regexparam.parse("/", true),
      modules: [
        {
          id: "_layout.tsx",
        },
      ],
    };

    const modules: RouteModules = {
      "_layout.tsx": {
        path: "src/_layout.tsx",
        load: async () => ({
          components: {
            main: () => <div>Layout</div>,
          },
        }),
      },
    };

    await loadRouteHandlers({ handlers, modules });

    expect(handlers).toMatchSnapshot();
  });

  test("1 route handler with 1 _index.tsx module", async () => {
    const handlers: RouteHandler = {
      route: {
        path: "/",
        mod: ["_index.tsx"],
        link: { css: [], mod: [] },
      },
      test: regexparam.parse("/", true),
      modules: [
        {
          id: "_index.tsx",
        },
      ],
    };

    const modules: RouteModules = {
      "_index.tsx": {
        path: "src/_index.tsx",
        load: async () => ({
          components: {
            main: () => <div>Index</div>,
          },
        }),
      },
    };

    await loadRouteHandlers({ handlers, modules });

    expect(handlers).toMatchSnapshot();
  });
});
