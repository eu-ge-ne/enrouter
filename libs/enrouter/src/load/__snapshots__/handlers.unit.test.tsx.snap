// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`loadRouteHandlers > 1 route handler with 1 _index.tsx module 1`] = `
{
  "index": {
    "main": {
      "$$typeof": Symbol(react.transitional.element),
      "_owner": null,
      "_store": {},
      "key": null,
      "props": {},
      "type": [Function],
    },
  },
  "modules": [
    {
      "id": "_index.tsx",
      "loaded": true,
    },
  ],
  "route": {
    "link": [
      [],
      [],
    ],
    "mod": [
      "_index.tsx",
    ],
    "path": "/",
  },
  "test": {
    "keys": [],
    "pattern": /\\^\\(\\?=\\$\\|\\\\/\\)/i,
  },
}
`;

exports[`loadRouteHandlers > 1 route handler with 1 _layout.tsx module 1`] = `
{
  "layout": {
    "main": {
      "$$typeof": Symbol(react.transitional.element),
      "_owner": null,
      "_store": {},
      "key": null,
      "props": {},
      "type": [Function],
    },
  },
  "modules": [
    {
      "id": "_layout.tsx",
      "loaded": true,
    },
  ],
  "route": {
    "link": [
      [],
      [],
    ],
    "mod": [
      "_layout.tsx",
    ],
    "path": "/",
  },
  "test": {
    "keys": [],
    "pattern": /\\^\\(\\?=\\$\\|\\\\/\\)/i,
  },
}
`;

exports[`loadRouteHandlers > 1 route handler with no modules 1`] = `
{
  "modules": [],
  "route": {
    "link": [
      [],
      [],
    ],
    "mod": [],
    "path": "/",
  },
  "test": {
    "keys": [],
    "pattern": /\\^\\(\\?=\\$\\|\\\\/\\)/i,
  },
}
`;

exports[`loadRouteHandlers > 2 route handlers with no modules 1`] = `
{
  "children": [
    {
      "modules": [],
      "route": {
        "link": [
          [],
          [],
        ],
        "mod": [],
        "path": "/abc",
      },
      "test": {
        "keys": [],
        "pattern": /\\^\\\\/abc\\(\\?=\\$\\|\\\\/\\)/i,
      },
    },
  ],
  "modules": [],
  "route": {
    "link": [
      [],
      [],
    ],
    "mod": [],
    "path": "/",
  },
  "test": {
    "keys": [],
    "pattern": /\\^\\(\\?=\\$\\|\\\\/\\)/i,
  },
}
`;
